### PR TITLE
Collaboration Tools

### PR DESCRIPTION
**Pull request by maarzt**
This is currently a prototype to solve #42 #43

This PR adds new menu items in Mastodon, that allow to share a Mastodon dataset, and to collaborate on it.
image

### TODOs
- [ ] ~Support SSH key on windows~
- [x] Better titles for the commands
- [x] Clone & Create repository commands, have a checkbox "create subdirectory"
- [x] Move clone command to Fiji menu
- [x] Rename menu entries
- [x] Test what happens if there is no global .gitconfig 
- [x] Tell users if credentials are wrong
- [x] Clone dialog: should have an example url
- [x] Change menu entries:
  - Push -> Upload changes,
  - Pull -> Download changes,
  - Create New Repository -> Share Project
- [x] Fix bug: Switch to remote branch does not work because “git fetch” is missing
- [x] Fix bug: Merge fails if dataset.xml is not there
- [x] Make commands cancellable
- [x] Warn user if author name was not specified yet
- [x] Disable commands that are not available outside of git repo
- [x] Allow to perform a pull if there are local and remote changes, provide options:
   1. delete all local changes
   2. ~create a branch to store local changes~
   3. cancel
- [x] write documentation

### CODE REVIEW

**Comment by maarzt** [comment]
@stefanhahmann Could you please test if the Mastodon git integration works:
* Clone this test repo: https://github.com/maarzt/mgit-test
* Make a change & commit
* Push the commit

I'm looking forward to get your feedback.

**Review by stefanhahmann — CHANGES_REQUESTED** [review]
This PR adds a bunch of functionalities needed for the git approach to Mastodon collaborative. The added features seem to be very useful.

It is a bit unfortunate that the test coverage for the logic contained in it is so low. This would make it hard to make changes in this repository for new developers in the future. You may consider to increase test coverage.

Apart from that I have mainly some comments on some terms shown to the user on the UI side and the user feedback on some specific workflows.

File: src/main/java/org/mastodon/mamut/tomancak/collaboration/credentials/PersistentCredentials.java [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+import org.eclipse.jgit.transport.URIish;
+
+public class PersistentCredentials
+{
+
+	private String username = null;
+
+	private String password = null;
+
+	private boolean missingCredentials()
```
[the remark this answers is not in the document]
Comment (maarzt) [reply src/main/java/org/mastodon/mamut/tomancak/collaboration/credentials/PersistentCredentials.java]: Addressed in 1bec3102ac2ff49e279bb779e1a96919f66420aa

File: src/main/java/org/mastodon/mamut/tomancak/collaboration/credentials/PersistentCredentials.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+import javax.swing.JTextField;
+
+import net.miginfocom.swing.MigLayout;
+
+import org.eclipse.jgit.errors.UnsupportedCredentialItem;
+import org.eclipse.jgit.transport.CredentialItem;
+import org.eclipse.jgit.transport.CredentialsProvider;
+import org.eclipse.jgit.transport.URIish;
+
+public class PersistentCredentials
```
[the remark this answers is not in the document]
Comment (masgitoff) [reply src/main/java/org/mastodon/mamut/tomancak/collaboration/credentials/PersistentCredentials.java:32]: Fixed in commit 97c225e52667a9f43ff76bf6931e8f1d77c76748

[the remark this answers is not in the document]
Comment (maarzt) [reply src/main/java/org/mastodon/mamut/tomancak/collaboration/credentials/PersistentCredentials.java:32]: Fixed in commit https://github.com/mastodon-sc/mastodon-tomancak/commit/6d773fdcd0d309172c73048b1b42d3a36abed015

File: src/main/java/org/mastodon/mamut/tomancak/collaboration/commands/MastodonGitCloneRepository.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+	{
+		try
+		{
+			directory = NewDirectoryUtils.createRepositoryDirectory( createSubdirectory, directory, repositoryURL );
+			MastodonGitRepository.cloneRepository( repositoryURL, directory );
+			MastodonGitRepository.openProjectInRepository( context, directory );
+		}
+		catch ( final Exception e )
+		{
+			ErrorDialog.showErrorMessage( "Download Shares Project (Clone)", e );
```
[the remark this answers is not in the document]
Comment (maarzt) [reply src/main/java/org/mastodon/mamut/tomancak/collaboration/commands/MastodonGitCloneRepository.java:63]: fixed

File: src/main/java/org/mastodon/mamut/tomancak/collaboration/commands/NewDirectoryUtils.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+		return createSubdirectory( directory, extractRepositoryName( repositoryURL ) );
+	}
+
+	static File createSubdirectory( File parentDirectory, String repositoryName ) throws IOException
+	{
+		File directory = new File( parentDirectory, repositoryName );
+		if ( directory.isDirectory() )
+		{
+			if ( directory.list().length > 0 )
```
[the remark this answers is not in the document]
Comment (maarzt) [reply src/main/java/org/mastodon/mamut/tomancak/collaboration/commands/NewDirectoryUtils.java:60]: I get this error message in my IDE too. But it's a false positive as it points to an non existing problem. The File.list() returns null only if the file is not a directory, but in line 58 I make sure that it is in fact a directory.

File: src/main/java/org/mastodon/mamut/tomancak/collaboration/commands/NewDirectoryUtils.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+				throw new IOException( "Directory already exists and is not empty: " + directory );
```
[the remark this answers is not in the document]
Comment (maarzt) [reply src/main/java/org/mastodon/mamut/tomancak/collaboration/commands/NewDirectoryUtils.java:61]: done

File: src/main/java/org/mastodon/mamut/tomancak/collaboration/commands/MastodonGitCloneRepository.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+		}
```
[the remark this answers is not in the document]
Comment (maarzt) [reply src/main/java/org/mastodon/mamut/tomancak/collaboration/commands/MastodonGitCloneRepository.java:64]: Fixed in 050375b43acbf4d5ed5cdc5a34674b59eba970e0

The stack trace is no longer written to the console. It instead can be shown by clicking on details in the error message dialog:
image

File: src/main/java/org/mastodon/mamut/tomancak/collaboration/commands/NewDirectoryUtils.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (stefanhahmann) [reply src/main/java/org/mastodon/mamut/tomancak/collaboration/commands/NewDirectoryUtils.java]: I disagree that this was a false positive.

According to documentation, there are two reasons why `File.list()` may return `null`.

- File is not a directory (which you checked)
- There as an I/O error while reading the directory (probably a rare case, but it was not checked)

Cf.:
grafik

File: src/main/java/org/mastodon/mamut/tomancak/collaboration/utils/ConflictUtils.java [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+		return name.equals( "Merge Conflict" ) ||
+				name.equals( "Merge Conflict (Tags)" ) ||
+				name.equals( "Merge Conflict (Labels)" ) ||
+				name.equals( "Merge Source A" ) ||
+				name.equals( "Merge Source B" ) ||
+				name.startsWith( "((A)) " ) ||
+				name.startsWith( "((B)) " );
+	}
+
+	private static boolean isTagSetEmpty( TagSetModel< Spot, Link > tagSetModel, String tagSetName, String tagLabel )
```
[the remark this answers is not in the document]
Comment (maarzt) [reply src/main/java/org/mastodon/mamut/tomancak/collaboration/utils/ConflictUtils.java:49]: Hm. The behavior of the class is "Returns true if the tag set is empty or if it does not exist."
I would keep this method here like that.
A method in TagSetUtils should throw a NoSuchElement exception if the tag set does not exist.

File: src/main/java/org/mastodon/mamut/tomancak/collaboration/ErrorDialog.java [hunk 8]
```diff
@@ -0,0 +1,20 @@
+package org.mastodon.mamut.tomancak.collaboration;
+
+import javax.swing.JOptionPane;
+import javax.swing.SwingUtilities;
+
+public class ErrorDialog
+{
+	public static void showErrorMessage( String title, Exception e )
+	{
+		e.printStackTrace();
```
[the remark this answers is not in the document]
Comment (maarzt) [reply src/main/java/org/mastodon/mamut/tomancak/collaboration/ErrorDialog.java]: I guess it's the most commonly accepted way to print a strack trace in scijava / Fiji.

File: src/main/java/org/mastodon/mamut/tomancak/collaboration/MastodonGitController.java [hunk 9]
```diff
@@ -0,0 +1,407 @@
...
+	private CommandService commandService;
+
+	@Parameter
+	private MastodonGitSettingsService settingsService;
+
+	public static final ActionDescriptions< MastodonGitController > actionDescriptions = new ActionDescriptions<>( MastodonGitController.class );
+
+	private static final String SHARE_PROJECT_ACTION_KEY = actionDescriptions.addActionDescription(
+			"[mastodon git] share project",
+			"Plugins > Git > Initialize > Share Project",
```
[the remark this answers is not in the document]
Comment (stefanhahmann) [reply src/main/java/org/mastodon/mamut/tomancak/collaboration/MastodonGitController.java:71]: I know that "Plugins > Collaborative" exists. This is why I suggested "Plugins > Collaborative **_(Git)_**"
I would argue that this name is more self-explaining to less technically people trained people, while it is still distinctive to the already existing Collaborative tools.

[the remark this answers is not in the document]
Comment (maarzt) [reply src/main/java/org/mastodon/mamut/tomancak/collaboration/MastodonGitController.java:71]: I agree with you. I even renamed the entry in the Fiji menu to "Plugins > Mastodon Collaborative (Git)". So it makes sense to rename it here as well.

**Comment by maarzt** [comment]
The classes in the PR where moved to a new repository https://github.com/maarzt/mastodon-git

**Pull request closed without merging**